local ReplicatedStorage = game:GetService("ReplicatedStorage")

local Common = ReplicatedStorage.Common

local Skins = require(Common.Skins.Skins)
export type SkinInfo = Skins.SkinInfo

local Various = {
    Skins.Grass,
    Skins.Wood,
    Skins.Stone,
    Skins.Brick,
    Skins.Rock,
    Skins.Obsidian,
    Skins.Amethyst,
    Skins.Magma,
    Skins.Lava,
    Skins.GoldOre,
    Skins.DiamondOre,
    Skins.GalaxyOre,
    Skins.Diamonds,
    Skins.Space,
    Skins.Rainbow,
    Skins.MemeDog,
    Skins.MemeCat,
    Skins.HungryCat,
    Skins.CuteCat,
    Skins.BananaCat,
    Skins.Backrooms,
    Skins.Water,
    Skins.Blank,
    Skins.Sahur,
    Skins.Trallalero,
    Skins.Cappuccino,
    Skins.Robux,
    Skins.Dollar,
    Skins.Warning,
    Skins.Cards,
    Skins.Chess,
    Skins.Stars,
    Skins.Squid,
    Skins.Doge,
    Skins.Labubu,
    Skins.Troll,
    Skins.SmileDog,
    Skins.FluffyDog,
    Skins.DripDog,
}

local Flags = {
    Skins.USA,
    Skins.China,
    Skins.Brazil,
    Skins.Philippines,
    Skins.Mexico,
    Skins.Japan,
    Skins.SouthKorea,
    Skins.Russia,
    Skins.Germany,
    Skins.France,
    Skins.UK,
    Skins.India,
    Skins.Indonesia,
    Skins.Malaysia,
    Skins.Thailand,
    Skins.Canada,
    Skins.Italy,
    Skins.Spain,
    Skins.Turkey,
    Skins.Australia,
    Skins.Poland,
    Skins.Netherlands,
    Skins.Argentina,
    Skins.Vietnam,
    Skins.Ukraine,
    Skins.UAE,
    Skins.Singapore,
    Skins.Switzerland,
    Skins.Sweden,
    Skins.Pakistan,
    Skins.Egypt,
    Skins.Colombia,
    Skins.Romania,
    Skins.Peru,
    Skins.Chile,
    Skins.Bangladesh,
    Skins.Belgium,
    Skins.Portugal,
    Skins.Austria,
    Skins.Czechia,
    Skins.Norway,
    Skins.Denmark,
    Skins.Finland,
    Skins.Jamaica,
    Skins.Ireland,
    Skins.Serbia,
    Skins.NewZealand,
    Skins.Greece,
    Skins.Scotland,
    Skins.Ghana,
    Skins.Kenya,
}

local SkinShop = {
    Various = Various,
    Flags = Flags,
}

return SkinShop